server/sonar-process/src/main/java/org/sonar/process/jmvoptions/JvmOptions.java

```diff
@@ -55,14 +114,37 @@ public T addFromMandatoryProperty(Props props, String propertyName) {
   public T add(String str) {
     requireNonNull(str, JVM_OPTION_NOT_NULL_ERROR_MESSAGE);
     String value = str.trim();
-    if (value.isEmpty() || !value.startsWith("-")) {
+    if (isInvalidOption(value)) {
       throw new IllegalArgumentException("a JVM option can't be empty and must start with '-'");
     }
+    checkMandatoryOptionOverwrite(value);
     options.add(value);
 
     return castThis();
   }
 
+  private void checkMandatoryOptionOverwrite(String value) {
+    Map.Entry<String, String> overriddenMandatoryOption = mandatoryOptionFor(value);
+    if (overriddenMandatoryOption != null) {
+      throw new MessageException(String.format(
+        "a JVM option can't overwrite mandatory JVM options. %s overwrites %s",
+        value,
+        overriddenMandatoryOption.getKey() + overriddenMandatoryOption.getValue()));
+    }
+  }
+
+  @CheckForNull
+  private Map.Entry<String, String> mandatoryOptionFor(String jvmOption) {
+    return mandatoryOptions.entrySet().stream()
+      .filter(s -> jvmOption.startsWith(s.getKey()) && !jvmOption.equals(s.getKey() + s.getValue()))
```

Comment:
From my point of view, we should display a message when a user is setting the property in both place even with same value.
And I think we may have corner case here, let's image that the JVM allows two options :
- -a
- -ab

We are not allowing to add `-ab` to optional parameters if `-a` is a mandatory parameter.
So maybe the solution is to extract the key from the configuration property and compare the keys without the value.

Reply:
This is theoretical. I considered it, looked a bit into it and dismissed the case.

As per the documentation provided by Oracle, there is no such case that I could find.
In addition, as for the option we actually define in our code (which is a restricted set of the available options), there is no such case for sure.